packages/financial-templates-lib/src/clients/OptimisticOracleClient.js

```diff
@@ -73,54 +100,125 @@ class OptimisticOracleClient {
       this.oracle.methods.getCurrentTime().call()
     ]);
 
-    // Fetch request states in parallel batches, 20 at a time, to be safe and not overload the web3 node.
-    const WEB3_CALLS_BATCH_SIZE = 150;
-
-    const [requestStates] = await Promise.all([
-      Promise.map(
-        requestEvents,
-        request =>
-          // Since we're making an async web3 call here, consider calling
-          // requests(web3.utils.soliditySha3(requester, id, timestamp, acData)) instead
-          // so we can grab more data including that relevant for proposals.
-          this.oracle.methods
-            .getState(
-              request.returnValues.requester,
-              request.returnValues.identifier,
-              request.returnValues.timestamp,
-              request.returnValues.ancillaryData ? request.returnValues.ancillaryData : "0x"
+    // Store price requests that have NOT been proposed to yet:
+    const unproposedPriceRequests = requestEvents.filter(event => {
+      const key = this._getPriceRequestKey(event);
+      const hasProposal = proposalEvents.find(proposalEvent => this._getPriceRequestKey(proposalEvent) === key);
+      return hasProposal === undefined;
+    });
+    this.unproposedPriceRequests = unproposedPriceRequests.map(event => {
+      return {
+        requester: event.returnValues.requester,
+        identifier: this.hexToUtf8(event.returnValues.identifier),
+        timestamp: event.returnValues.timestamp,
+        currency: event.returnValues.currency,
+        reward: event.returnValues.reward,
+        finalFee: event.returnValues.finalFee
+      };
+    });
+
+    // Store proposals that have NOT been disputed:
+    let undisputedProposals = proposalEvents.filter(event => {
+      const key = this._getPriceRequestKey(event);
+      const hasDispute = disputeEvents.find(disputeEvent => this._getPriceRequestKey(disputeEvent) === key);
+      return hasDispute === undefined;
+    });
+    undisputedProposals = undisputedProposals.map(event => {
+      return {
+        requester: event.returnValues.requester,
+        proposer: event.returnValues.proposer,
+        identifier: this.hexToUtf8(event.returnValues.identifier),
+        timestamp: event.returnValues.timestamp,
+        proposedPrice: event.returnValues.proposedPrice,
+        expirationTimestamp: event.returnValues.expirationTimestamp
+      };
+    });
+
+    // Filter proposals based on their expiration timestamp:
+    const isExpired = proposal => {
+      return Number(proposal.expirationTimestamp) <= Number(currentTime);
+    };
+    this.expiredProposals = undisputedProposals.filter(proposal => isExpired(proposal));
+    this.undisputedProposals = undisputedProposals.filter(proposal => !isExpired(proposal));
+
+    // Store disputes that can be settled:
+    let resolvedPrices = await Promise.all(
+      disputeEvents.map(async disputeEvent => {
+        try {
```

Comment:
do we want to have this `try` nested within a `Promise.all, within a `map`? is there a way to change this to remove the nesting a bit and deal with the error the try is catching a bit sooner?

Reply:
What we could do is `map` the `try-catch`, and then pass the map of promises into a `Promise.all`. I wanted to be more concise, but I understand if readability is decreased a lot.